### PR TITLE
refactor: hide format deserializer state type

### PR DESCRIPTION
**Pull request by leostera**
**tldr:** go from `(Serde_json.De.state, User.t) Serde.De.t` to `User.t Serde.De.t` 

This refactor should make the ergonomics of Serde much better for library authors that need to provide flexibility on the format deserializer type, by making the format deserializer state type _go away_ or be hidden at least from the user facing code.

We've first observed this when working on DbCaml with @emilpriver and noticed that a common error was unifying the types of the state of the format deserializer, which wasn't known in some cases by design, to allow a user to switch up which database format deserializer module they wanted to use (Postgres, SQLite, etc).

### CODE REVIEW

**Comment by leostera** [comment]
Confirmed with the bug repro repo (https://github.com/leostera/serde-de-state-bug-repr/commit/174d2409f93e1186fc5dded34babe8bcde7d6c74) and with @emil_priver that this works for DbCaml too.

Merging 🚀
